website/docs/language/expressions/custom-conditions.mdx

```diff
@@ -0,0 +1,353 @@
+```
+
+```hcl
+  # This remote output value must be convertible to a list
+  # type of with element type.
+  condition = can(tolist(data.terraform_remote_state.example.outputs["items"]))
+```
+
+You can also use `can` with attribute access or index operators to test whether a collection or structural value has a particular element or index.
+
+```hcl
+  # var.example must have an attribute named "foo"
+  condition = can(var.example.foo)  ```
+
+```hcl
+  # var.example must be a sequence with at least one element
+  condition = can(var.example[0])
+  # (although it would typically be clearer to write this as a
+  # test like length(var.example) > 0 to better represent the
+  # intent of the condition.)
+```
+
+### `self` Object
+
+Use the `self` object in postcondition blocks to refer to attributes of the instance under evaluation.
+
+```hcl
+resource "aws_instance" "example" {
+  instance_type = "t2.micro"
+  ami           = "ami-abc123"
+
+  lifecycle {
+    postcondition {
+      condition     = self.instance_state == "running"
+      error_message = "EC2 instance must be running."
+    }
+  }
+}
+```
+
+### `each` and `count` Objects
+
+In blocks where `for_each` or `count` are set, use `each` and `count` objects to refer to other resources that are expanded in a chain.
+
+```hcl
+variable "vpc_cidrs" {
+  type = set(string)
+}
+
+data "aws_vpc" "example" {
+  for_each = var.vpc_cidrs
+
+  filter {
+    name   = "cidr"
+    values = [each.key]
+  }
+}
+
+resource "aws_internet_gateway" "example" {
+  for_each = aws_vpc.example
+  vpc_id = each.value.id
+
+  lifecycle {
+    precondition {
+      condition     = aws_vpc.example[each.key].state == "available"
+      error_message = "VPC ${each.key} must be available."
+    }
+  }
+}
+```
+
+## Error Messages
+
+Input variable validations, preconditions, and postconditions all must include the `error_message` argument. This contains the text that Terraform will include as part of error messages when it detects an unmet condition.
+
+```
+Error: Resource postcondition failed
+
+  with data.aws_ami.example,
+  on ec2.tf line 19, in data "aws_ami" "example":
+  72:       condition     = self.tags["Component"] == "nomad-server"
+    |----------------
+    | self.tags["Component"] is "consul-server"
+
+The selected AMI must be tagged with the Component value "nomad-server".
+```
+
+The `error_message` argument can be any expression that evaluates to a string.
+This includes literal strings, heredocs, and template expressions. Multi-line
+error messages are supported, and lines with leading whitespace will not be
+word wrapped.
+
+We recommend writing error messages as one or more full sentences in a
+style similar to Terraform's own error messages. Terraform will show the
+message alongside the name of the resource that detected the problem and any
+external values included in the condition expression.
+
+## Delaying Evaluation Until Apply
```

Comment:
I feel a bit torn about this section in particular because it kinda feels like it's a description of how to _intentionally_ delay evaluation until apply, but in practice checking things during apply is the worst case and the guidance here is, I think, more about how to _avoid_ delaying evaluation until apply. (It's typically desirable to get error message feedback as at early a phase as possible, and always better to see an error before apply if possible.)

With that said though, a heading like "Checking Conditions Before Apply" doesn't really feel right either. I think what makes this tricky is that this section is here to explain something a user might not even think to ask, and so it's hard to imagine what sort of heading would cause them to see this as important information to learn about.

One compromise that I can think of is to include an info callout :information_source: just before the current "`contains` Function" heading above, at the end of the paragraphs under "Condition Expressions", with a short note something like this:

> **Note:** Terraform will check conditions during the planning step when possible, but some condition expressions can be checked only during the apply process. For more information, see <i>Conditions Checked Only During Apply</i>.

(where that last part "Conditions Checked Only During Apply" is a link to the section where I've attached this comment, which I guess I'm also implicitly proposing to rename to "Conditions Checked Only During Apply".)

Reply:
Okay first - this is ANOTHER thing I struggled with, so I'm really glad to have your thoughts. Do we think that is an okay title? I'm going to change it to that (I wonder to be honest if I meant to and then in my reorganization frenzy did not change the main title. Thank you!! I'm going to change it to that for now, and then I think maybe I'll also think about it some more.